In [2]:
import pandas as pd


df_row = pd.read_csv("FoodFacts.products_Dataset.170k_May26.csv")



/tmp/ipykernel_971/1365844784.py:4: DtypeWarning: Columns (20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,51,52,53,54,55,56,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272,273,274,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,304,305,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,

In [3]:
print("Shape:", df_row.shape)

print("\nColumns:")
print(df_row.columns.tolist())

print("\nTarget Distribution:")
print(df_row["ecoscore_grade"].value_counts(dropna=False))

Shape: (10692, 509)

Columns:
['_id', 'product_name_en', 'ecoscore_grade', 'ingredients_text_en', 'categories', 'categories_hierarchy[0]', 'categories_hierarchy[1]', 'categories_hierarchy[2]', 'categories_hierarchy[3]', 'categories_hierarchy[4]', 'categories_hierarchy[5]', 'categories_hierarchy[6]', 'categories_hierarchy[7]', 'categories_hierarchy[8]', 'categories_hierarchy[9]', 'categories_hierarchy[10]', 'categories_hierarchy[11]', 'categories_hierarchy[12]', 'categories_hierarchy[13]', 'categories_hierarchy[14]', 'categories_hierarchy[15]', 'categories_hierarchy[16]', 'categories_hierarchy[17]', 'categories_hierarchy[18]', 'categories_hierarchy[19]', 'categories_hierarchy[20]', 'categories_hierarchy[21]', 'categories_hierarchy[22]', 'categories_hierarchy[23]', 'categories_hierarchy[24]', 'categories_hierarchy[25]', 'categories_hierarchy[26]', 'categories_hierarchy[27]', 'categories_hierarchy[28]', 'categories_hierarchy[29]', 'categories_hierarchy[30]', 'categories_hierarchy[31]', 'c

In [4]:
missing = (
    df_row.isnull().sum()
    .sort_values(ascending=False)
)

print(missing.head(30))

packaging_hierarchy[46]     10692
packaging_hierarchy[47]     10692
packaging_hierarchy[48]     10692
categories_hierarchy[37]    10692
categories_hierarchy[38]    10692
categories_hierarchy[39]    10692
categories_hierarchy[40]    10692
packaging_hierarchy[50]     10692
packaging_hierarchy[32]     10692
packaging_hierarchy[33]     10692
packaging_hierarchy[34]     10692
packaging_hierarchy[35]     10692
packaging_hierarchy[36]     10692
packaging_hierarchy[37]     10692
packaging_hierarchy[38]     10692
packaging_hierarchy[39]     10692
packaging_hierarchy[40]     10692
packaging_hierarchy[41]     10692
packaging_hierarchy[42]     10692
countries_hierarchy[10]     10692
countries_hierarchy[11]     10692
countries_hierarchy[12]     10692
countries_hierarchy[13]     10692
categories_hierarchy[32]    10692
categories_hierarchy[33]    10692
packaging_hierarchy[15]     10692
packaging_hierarchy[16]     10692
packaging_hierarchy[17]     10692
packaging_hierarchy[18]     10692
countries_hier

In [5]:
print(df_row['categories'].head(10))

0    Beverages and beverages preparations, Beverage...
1    Dranken en drankbereidingen,Dranken,Alcoholisc...
2                                             Biscuits
3                                      Green tapenades
4                                     Chocolate Cookie
5    Alimentos y bebidas de origen vegetal, Aliment...
6    Plant-based foods and beverages, Plant-based f...
7    Dairies, Desserts, Dairy desserts, fr:cremes-b...
8    Plant-based foods and beverages, Plant-based f...
9    Plant-based foods and beverages, Plant-based f...
Name: categories, dtype: object


In [6]:
print(df_row['categories'].sample(10, random_state=42))

1964    Dairies, Fermented foods, Fermented milk produ...
5267    Plant-based foods and beverages, Beverages, Pl...
9630                Beverages, Alcoholic beverages, Beers
5503    Plant-based foods and beverages, Plant-based f...
1683                                       Dairies, Milks
1983    Snacks, Sweet snacks, Biscuits and cakes, Bisc...
1056    Plant-based foods and beverages, Beverages, Pl...
6224        Beverages, Alcoholic beverages, Beers, Lagers
1732                                       Dairies, Milks
7934    Plant-based foods and beverages, Plant-based f...
Name: categories, dtype: object


In [7]:
df_row['categories'].isnull().sum()

np.int64(0)

In [8]:
ingredient_cols = [
    c for c in df_row.columns
    if 'ingredients_hierarchy[' in c
]

In [9]:
category_cols = [
    c for c in df_row.columns
    if 'categories_hierarchy[' in c
]

In [10]:
print(len(ingredient_cols))
print(len(category_cols))

365
42


In [11]:
def merge_hierarchy(row):

    values = []

    for item in row:

        if pd.notnull(item):

            item = str(item)

            item = item.replace('en:', '')
            item = item.replace('-', ' ')
            item = item.strip()

            if item:
                values.append(item)

    values = list(dict.fromkeys(values))

    return ' '.join(values)

In [12]:
df_row['ingredient_hierarchy_text'] = (
    df_row[ingredient_cols]
    .apply(merge_hierarchy, axis=1)
)
df_row['category_hierarchy_text'] = (
    df_row[category_cols]
    .apply(merge_hierarchy, axis=1)
)

In [13]:
print(df_row['ingredient_hierarchy_text'].head())
print("*"*50)
print(df_row['category_hierarchy_text'].head())

0    sugar added sugar disaccharide whole milk dair...
1                                                     
2                                                     
3                                                     
4                                                  dog
Name: ingredient_hierarchy_text, dtype: object
**************************************************
0    beverages and beverages preparations plant bas...
1    beverages and beverages preparations beverages...
2      snacks sweet snacks biscuits and cakes biscuits
3     spreads salted spreads tapenades green tapenades
4    snacks sweet snacks biscuits and cakes biscuit...
Name: category_hierarchy_text, dtype: object


In [14]:
print(
    df_row[
        [
            'product_name_en',
            'ingredients_text_en',
            'categories',
            'ingredient_hierarchy_text',
            'category_hierarchy_text'
        ]
    ].isnull().sum()
)

product_name_en                 0
ingredients_text_en          3839
categories                      0
ingredient_hierarchy_text       0
category_hierarchy_text         0
dtype: int64


In [15]:
(
    df_row['ingredient_hierarchy_text']
    .str.strip()
    .eq('')
    .sum()
)

np.int64(3619)

In [16]:
(
    df_row['category_hierarchy_text']
    .str.strip()
    .eq('')
    .sum()
)

np.int64(0)

In [17]:
df_row["_id"].value_counts()

,count
_id,
4.088600e+12,122
4.056490e+12,83
4.099100e+12,75
5.000170e+12,75
5.010250e+12,57
...,...
8.545900e+11,1
6.738030e+11,1
8.513610e+11,1


In [18]:
(
    (
        df_row['product_name_en'].fillna('').str.strip() == ''
    ) &
    (
        df_row['ingredients_text_en'].fillna('').str.strip() == ''
    ) &
    (
        df_row['categories'].fillna('').str.strip() == ''
    ) &
    (
        df_row['ingredient_hierarchy_text'].fillna('').str.strip() == ''
    ) &
    (
        df_row['category_hierarchy_text'].fillna('').str.strip() == ''
    )
).sum()

np.int64(0)

In [19]:
df_row[['_id', 'product_name_en']].head(20)

,_id,product_name_en
0,7.896010e+12,123
1,8.410390e+12,2022
2,8.901060e+12,5050
3,4.004200e+12,5.00017E+12
4,5.203060e+12,5.20306E+12
5,7.771260e+12,7.77126E+12
6,2.617318e+07,Asparagus Spears
7,2.527927e+07,Creme Brûlée
8,3.045320e+12,Peach Conserve
9,9.415080e+12,pitted dates


In [20]:
df_nlp = df_row[
    [
        'ecoscore_grade',
        'product_name_en',
        'ingredients_text_en',
        'categories',
        'ingredient_hierarchy_text',
        'category_hierarchy_text'
    ]
].copy()

In [21]:
df_nlp[
    df_nlp['ingredient_hierarchy_text']
    .astype(str)
    .str.fullmatch(r'e\d+[a-z]*', na=False)
].shape

(5, 6)

In [22]:
df_row[
    df_row['ingredient_hierarchy_text']
    .astype(str)
    .str.fullmatch(r'[\d\.Ee\+\-]+', na=False)
][['_id', 'ingredient_hierarchy_text']].head(20)

,_id,ingredient_hierarchy_text
1348,5.010510e+12,e150
2935,7.337391e+11,e406
3438,7.882692e+11,e123
3537,8.583200e+11,e967
9614,5.054070e+12,e500


In [23]:
(
    df_row["ingredient_hierarchy_text"]
    .fillna('')
    .astype(str)
    .str.fullmatch(r'[\d\.Ee\+\-]+')
    .sum()
)

np.int64(5)

In [24]:
df_row = df_row[
    ~(
        (
            df_row['product_name_en'].fillna('').str.strip() == ''
        ) &
        (
            df_row['ingredients_text_en'].fillna('').str.strip() == ''
        ) &
        (
            df_row['categories'].fillna('').str.strip() == ''
        ) &
        (
            df_row['ingredient_hierarchy_text'].fillna('').str.strip() == ''
        ) &
        (
            df_row['category_hierarchy_text'].fillna('').str.strip() == ''
        )
    )
].copy()

In [25]:
print(df_nlp.shape)

(10692, 6)


In [26]:
print(df_nlp.sample(5, random_state=42))

     ecoscore_grade                                    product_name_en  \
1964              d  365 everyday value, organic hickory smoked che...   
5267              c                                        Apple Juice   
9630              d                                      Biere de luxe   
5503              e            Apple Pieces & Cinnamon Instant Oatmeal   
1683              d                                2% reduced fat milk   

                                    ingredients_text_en  \
1964  Cultured pasteurized organic milk, salt, micro...   
5267                                                NaN   
9630                                                NaN   
5503                                                NaN   
1683  Reduced fat milk, vitamin a palmitate, vitamin...   

                                             categories  \
1964  Dairies, Fermented foods, Fermented milk produ...   
5267  Plant-based foods and beverages, Beverages, Pl...   
9630              Bever

In [27]:
df_nlp

,ecoscore_grade,product_name_en,ingredients_text_en,categories,ingredient_hierarchy_text,category_hierarchy_text
0,f,123,NaN,"Beverages and beverages preparations, Beverage...",sugar added sugar disaccharide whole milk dair...,beverages and beverages preparations plant bas...
1,e,2022,NaN,"Dranken en drankbereidingen,Dranken,Alcoholisc...",,beverages and beverages preparations beverages...
2,c,5050,NaN,Biscuits,,snacks sweet snacks biscuits and cakes biscuits
3,c,5.00017E+12,NaN,Green tapenades,,spreads salted spreads tapenades green tapenades
4,d,5.20306E+12,dog,Chocolate Cookie,dog,snacks sweet snacks biscuits and cakes biscuit...
...,...,...,...,...,...,...
10687,b,Blackberry jam,NaN,"Plant-based foods and beverages, Plant-based f...",,plant based foods and beverages plant based fo...
10688,b,Blackberry jam,NaN,"Plant-based foods and beverages, Plant-based f...",,plant based foods and beverages plant based fo...
10689,b,blackberry jam,NaN,"Plant-based foods and beverages, Plant-based f...",,plant based foods and beverages plant based fo...
10690,b,Blackberry jam,NaN,"Plant-based foods and beverages, Plant-based f...",,plant based foods and beverages plant based fo...


In [28]:
print(df_nlp['product_name_en'].sample(10, random_state=1).tolist())

['Apple Mix Big Pack', 'baked beans in tomato sauce', 'Azucar bakery, dulce de leche alfajor cookies', 'BBQ sauce mango-apricot', 'Beef lasagne', 'Bintang Bir Pilsener Beer', '4 Omega 3 Fish Fillets', '100% juice', '2% reduced fat milk', 'Baked Beans, Homestyle']


In [29]:
print(df_nlp['ingredients_text_en'].sample(10, random_state=1).tolist())

[nan, nan, 'Unbleached flour, cornstarch, powder sugar, milk, vegetable oil blend.', nan, 'morrisons since 1899 free from gluten free beef lasagne gluten free pasta layered with a rich bolognése sauce, topped with bechamel sauce and grated cheese 300ge a pack (microwave energy 1990 25.34 11.5', nan, 'Alaska Pollock Fillet (50%) (_Fish_), Wheat Flour (_Wheat_ Flour, Calcium Carbonate, Iron, Niacin (B3), Thiamin (B1)), Sunflower Oil, Rapeseed Oil, Palm Oil, Water, Salt, _Wheat_ Starch, Maize Flour, Raising Agents (Diphosphates, Sodium Bicarbonate), _Mustard_ Flour, Dextrose, _Wheat_ Gluten, Dried Skimmed _Milk_, Flavouring (contains _Mustard_), Colours (Curcumin, Capsanthin), Spices.', 'Water, grape juice concentrate, ascorbic acid (vitamin c).', 'Reduced fat milk, vitamin a palmitate and vitamin d3.', 'prepared white beans, water, brown sugar, sugar, tomato paste, bacon, salt, mustard (water, vinegar, mustard seed, salt, turmeric, spices), modified corn starch, onion powder, spices, ext

In [30]:
print(df_nlp['categories'].sample(10, random_state=1).tolist())

['en:candies', 'Plant-based foods and beverages, Plant-based foods, Prepared vegetables, Baked beans in tomato sauce', 'Snacks, Sweet snacks, Biscuits and cakes, Biscuits', 'Condiments, Sauces, Barbecue sauces', 'Meals, Pasta dishes, Prepared lasagne, Bolognese lasagne', 'Beverages, Alcoholic beverages, Beers', 'Seafood, Fishes, Fish fillets, Fish preparations, Breaded products, Breaded fish', 'Plant-based foods and beverages, Beverages, Plant-based beverages, Fruit-based beverages, Juices and nectars, Fruit juices, Fruit juices from concentrate, Grape juices, Grape juices from concentrate', 'Dairies, Milks', 'Plant-based foods and beverages, Plant-based foods, Meals, Prepared vegetables, Baked beans in tomato sauce']


In [31]:
print(df_nlp[['categories',
              'category_hierarchy_text']].sample(
                  10,
                  random_state=42
              ))

                                             categories  \
1964  Dairies, Fermented foods, Fermented milk produ...   
5267  Plant-based foods and beverages, Beverages, Pl...   
9630              Beverages, Alcoholic beverages, Beers   
5503  Plant-based foods and beverages, Plant-based f...   
1683                                     Dairies, Milks   
1983  Snacks, Sweet snacks, Biscuits and cakes, Bisc...   
1056  Plant-based foods and beverages, Beverages, Pl...   
6224      Beverages, Alcoholic beverages, Beers, Lagers   
1732                                     Dairies, Milks   
7934  Plant-based foods and beverages, Plant-based f...   

                                category_hierarchy_text  
1964  dairies fermented foods fermented milk product...  
5267  plant based foods and beverages beverages plan...  
9630                beverages alcoholic beverages beers  
5503  plant based foods and beverages plant based fo...  
1683                                      dairies milks  
19

In [32]:
df_nlp = df_row[
    [
        'ecoscore_grade',
        'product_name_en',
        'ingredients_text_en',
        'ingredient_hierarchy_text',
        'category_hierarchy_text'
    ]
].copy()

In [33]:
print(df_nlp.shape)

print(df_nlp.isnull().sum())

(10692, 5)
ecoscore_grade                  0
product_name_en                 0
ingredients_text_en          3839
ingredient_hierarchy_text       0
category_hierarchy_text         0
dtype: int64


In [34]:
(
    (
        df_nlp['product_name_en'].fillna('').str.strip() == ''
    ) &
    (
        df_nlp['ingredients_text_en'].fillna('').str.strip() == ''
    ) &
    (
        df_nlp['ingredient_hierarchy_text'].fillna('').str.strip() == ''
    ) &
    (
        df_nlp['category_hierarchy_text'].fillna('').str.strip() == ''
    )
).sum()

np.int64(0)

In [35]:
import re
import unicodedata

def clean_text_lstm(text):

    if pd.isna(text):
        return ''

    text = str(text)

    # unicode normalization
    text = unicodedata.normalize('NFKC', text)

    # remove language tags
    text = re.sub(
        r'\b(?:en|fr|de|pt|it|ru|es|nl):',
        ' ',
        text,
        flags=re.IGNORECASE
    )

    # remove urls
    text = re.sub(
        r'http\S+|www\S+',
        ' ',
        text
    )

    # replace hyphens
    text = text.replace('-', ' ')

    # remove punctuation
    text = re.sub(
        r'[^\w\s]',
        ' ',
        text
    )

    # normalize spaces
    text = re.sub(
        r'\s+',
        ' ',
        text
    ).strip()

    # lowercase
    text = text.lower()

    return text

In [36]:
text_cols = [
    'product_name_en',
    'ingredients_text_en',
    'ingredient_hierarchy_text',
    'category_hierarchy_text'
]

for col in text_cols:
    df_nlp[col] = (
        df_nlp[col]
        .fillna('')
        .apply(clean_text_lstm)
    )

In [37]:
print(df_nlp.sample(10, random_state=42))

     ecoscore_grade                                    product_name_en  \
1964              d  365 everyday value organic hickory smoked ched...   
5267              c                                        apple juice   
9630              d                                      biere de luxe   
5503              e              apple pieces cinnamon instant oatmeal   
1683              d                                 2 reduced fat milk   
1983              d              365 everyday value shortbread cookies   
1056              d                    100 pure juice from concentrate   
6224              c                                    asahi super dry   
1732              d                                 2 reduced fat milk   
7934              b                                         banana jam   

                                    ingredients_text_en  \
1964  cultured pasteurized organic milk salt microbi...   
5267                                                      
9630    

In [38]:
print(df_nlp.isnull().sum())

ecoscore_grade               0
product_name_en              0
ingredients_text_en          0
ingredient_hierarchy_text    0
category_hierarchy_text      0
dtype: int64


In [39]:
for col in [
    'product_name_en',
    'ingredients_text_en',
    'ingredient_hierarchy_text',
    'category_hierarchy_text'
]:

    non_empty = (
        df_nlp[col]
        .str.strip()
        .ne('')
        .sum()
    )

    print(col)
    print(non_empty)
    print(
        round(
            non_empty / len(df_nlp) * 100,
            2
        ),
        '%'
    )
    print()

product_name_en
10692
100.0 %

ingredients_text_en
6853
64.09 %

ingredient_hierarchy_text
7073
66.15 %

category_hierarchy_text
10692
100.0 %



In [40]:
df_nlp['deep_text'] = (
    df_nlp['product_name_en'] + ' ' +
    df_nlp['ingredients_text_en'] + ' ' +
    df_nlp['ingredient_hierarchy_text'] + ' ' +
    df_nlp['category_hierarchy_text']
)

In [41]:
print(df_nlp['deep_text'].sample(10, random_state=42))

1964    365 everyday value organic hickory smoked ched...
5267    apple juice   plant based foods and beverages ...
9630    biere de luxe   beverages alcoholic beverages ...
5503    apple pieces cinnamon instant oatmeal   plant ...
1683    2 reduced fat milk reduced fat milk vitamin a ...
1983    365 everyday value shortbread cookies unbleach...
1056    100 pure juice from concentrate water orange j...
6224    asahi super dry water _barley malt_ corn starc...
1732    2 reduced fat milk grade a reduced fat organic...
7934    banana jam   plant based foods and beverages p...
Name: deep_text, dtype: object


In [42]:
df_nlp['deep_text'].str.split().apply(len).describe()

,deep_text
count,10692.000000
mean,57.813505
std,60.123612
min,3.000000
25%,20.000000
50%,34.000000
75%,76.000000
max,1447.000000


In [43]:
lengths = (
    df_nlp['deep_text']
    .str.split()
    .apply(len)
)

print("90% :", lengths.quantile(0.90))
print("95% :", lengths.quantile(0.95))
print("99% :", lengths.quantile(0.99))

90% : 135.0
95% : 178.0
99% : 267.0


In [44]:
df_nlp.loc[
    lengths > 500,
    'deep_text'
].head(5)

,deep_text
4682,angus beef ribeye steak ribeye steak nutrition...
4808,antipasto misti prosciutto pork salt coppa di ...
4849,apetit αρετιt posni keks srb mne hr bih keks s...
6455,assorted cookies roscayema wheat flour sugar v...
8938,beef enchilada dinner tortilla corn flour grou...


In [45]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

In [46]:
label_encoder = LabelEncoder()

df_nlp['label'] = label_encoder.fit_transform(
    df_nlp['ecoscore_grade']
)

print(dict(zip(label_encoder.classes_,label_encoder.transform(   label_encoder.classes_) )))

{'a': np.int64(0), 'a-plus': np.int64(1), 'b': np.int64(2), 'c': np.int64(3), 'd': np.int64(4), 'e': np.int64(5), 'f': np.int64(6)}


In [47]:
X = df_nlp['deep_text']

y = df_nlp['label']

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [48]:
print(X_train.shape)
print(X_test.shape)

print(y_train.value_counts())

(8553,)
(2139,)
label
4    2437
2    1839
3    1391
0    1187
5    1084
6     472
1     143
Name: count, dtype: int64


### Text Vectorization: Bag-of-Words

We will use `CountVectorizer` to convert text data into numerical feature vectors using the Bag-of-Words (BoW) model. This approach counts the occurrences of words within a document, creating a vector where each dimension corresponds to a unique word in the vocabulary.

`max_features` controls the size of the vocabulary, `ngram_range` allows capturing combinations of words (bigrams in this case), and `min_df` filters out words that appear too infrequently.

In [49]:
from sklearn.feature_extraction.text import CountVectorizer

# Bag of Words (BoW) Vectorization
# max_features: Limits the vocabulary size to the most frequent words.
# ngram_range: Includes unigrams and bigrams for better context capture.
# min_df: Ignores terms that appear in less than 2 documents to reduce noise.
bow_vectorizer = CountVectorizer(
    max_features=100000,
    ngram_range=(1,2),
    min_df=2
)

# Fit the vectorizer on the training data and transform both training and test data
X_train_bow = bow_vectorizer.fit_transform(X_train)
X_test_bow = bow_vectorizer.transform(X_test)


### Handling Imbalanced Data with SMOTE

Our target variable `ecoscore_grade` might have an imbalanced distribution, which can negatively impact model performance. To address this, we use **SMOTE (Synthetic Minority Over-sampling Technique)**.

SMOTE works by creating synthetic samples from the minority class, helping to balance the class distribution. This prevents models from being biased towards the majority class and improves their ability to classify minority classes correctly.

In [50]:
from imblearn.over_sampling import SMOTE

# Initialize SMOTE with a fixed random state for reproducibility
smote = SMOTE(random_state=42)

# Apply SMOTE to the Bag-of-Words training data
# This will generate synthetic samples for minority classes to balance the dataset
X_train_bow_smote, y_train_bow_smote = smote.fit_resample(
    X_train_bow,
    y_train
)

print("Shape of X_train_bow after SMOTE:", X_train_bow_smote.shape)
print("Distribution of y_train_bow after SMOTE:")
print(pd.Series(y_train_bow_smote).value_counts())


Shape of X_train_bow after SMOTE: (17059, 28354)
Distribution of y_train_bow after SMOTE:
label
4    2437
2    2437
0    2437
3    2437
6    2437
5    2437
1    2437
Name: count, dtype: int64


### Text Classification Models

This section defines a class `TextClassificationModels` to train and evaluate various machine learning models. As requested, we will focus on a subset of models: **Naive Bayes**, **Random Forest**, and **XGBoost**. These models are well-suited for text classification tasks.

-   **Naive Bayes (MultinomialNB)**: A probabilistic classifier particularly effective for text data, assuming feature independence.
-   **Random Forest (RandomForestClassifier)**: An ensemble learning method that builds multiple decision trees and merges their predictions for improved accuracy and robustness.
-   **XGBoost (XGBClassifier)**: A powerful gradient boosting framework known for its speed and performance, especially on structured and tabular data, which can also be applied to text features.

Each model is trained on the preprocessed training data (`X_train_bow_smote`, `y_train_bow_smote`) and evaluated on the test set (`X_test_bow`, `y_test`) using accuracy and a detailed classification report.

In [51]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

from sklearn.metrics import classification_report, accuracy_score


class TextClassificationModels:

    def __init__(self):
        self.models = {
            "Naive Bayes": MultinomialNB(),
            "Random Forest": RandomForestClassifier(
                n_estimators=50,
                class_weight="balanced",
                random_state=42,
                n_jobs=-1 # Use all available cores for faster training
            ),
            "XGBoost": XGBClassifier(
                n_estimators=100,
                max_depth=8,
                learning_rate=0.1,
                objective='multi:softmax',
                num_class=7,
                random_state=42
            )
        }

    def train_and_evaluate(self, X_train, X_test, y_train, y_test, label_encoder):

        results = {}

        for name, model in self.models.items():
            print(f"\nTraining {name}...")

            # Fit the model on the training data
            model.fit(X_train, y_train)

            # Make predictions on the test set
            y_pred = model.predict(X_test)

            # Calculate accuracy
            acc = accuracy_score(y_test, y_pred)

            print(f"\n{name} Accuracy: {acc:.4f}")
            # Print detailed classification report
            print(
                classification_report(
                    y_test,
                    y_pred,
                    target_names=label_encoder.classes_
                )
            )

            results[name] = acc

        return results

In [52]:
ml_models = TextClassificationModels()

# --- Run models with original Bag-of-Words data (without SMOTE) ---
print("\n--- Training with original Bag-of-Words data ---")
bow_results = ml_models.train_and_evaluate(
    X_train_bow,
    X_test_bow,
    y_train,
    y_test,
    label_encoder
)

# --- Run models with Bag-of-Words data + SMOTE ---
print("\n--- Training with Bag-of-Words data + SMOTE ---")
bow_smote_results = ml_models.train_and_evaluate(
    X_train_bow_smote,
    X_test_bow,
    y_train_bow_smote,
    y_test,
    label_encoder
)

print("\nComparison of results:")
print("Original BoW results:", bow_results)
print("BoW + SMOTE results:", bow_smote_results)


--- Training with original Bag-of-Words data ---

Training Naive Bayes...

Naive Bayes Accuracy: 0.7162
              precision    recall  f1-score   support

           a       0.67      0.81      0.74       297
      a-plus       0.00      0.00      0.00        36
           b       0.76      0.68      0.72       460
           c       0.65      0.66      0.66       348
           d       0.79      0.72      0.76       609
           e       0.63      0.81      0.71       271
           f       0.76      0.75      0.75       118

    accuracy                           0.72      2139
   macro avg       0.61      0.63      0.62      2139
weighted avg       0.71      0.72      0.71      2139


Training Random Forest...

Random Forest Accuracy: 0.8144
              precision    recall  f1-score   support

           a       0.84      0.79      0.81       297
      a-plus       0.31      0.11      0.16        36
           b       0.81      0.84      0.83       460
           c       0.8

### Conclusion on Model Performance

Based on the evaluation of the three models (Naive Bayes, Random Forest, and XGBoost) with and without SMOTE for handling class imbalance, here's a summary of the results:

1.  **Overall Best Performer: XGBoost**
    *   XGBoost consistently achieved the highest accuracy in both scenarios. With the original Bag-of-Words (BoW) data, it scored 0.8266, and with SMOTE-resampled data, its accuracy slightly improved to 0.8308.
    *   This indicates that XGBoost is robust and effective for this text classification task, even with the high-dimensional sparse data from BoW vectorization.

2.  **Random Forest Performance**
    *   Random Forest also performed strongly, with an accuracy of 0.8144 on the original BoW data. Its performance slightly decreased to 0.8130 after applying SMOTE.
    *   While competitive, it was slightly outpaced by XGBoost.

3.  **Naive Bayes Performance**
    *   Naive Bayes showed a decent accuracy of 0.7162 on the original BoW data. However, its performance dropped to 0.6928 when trained on the SMOTE-resampled data.
    *   This suggests that SMOTE, in this particular application, might have introduced noise or synthetic patterns that did not benefit the Naive Bayes model as much, possibly due to its strong assumption of feature independence.

4.  **Impact of SMOTE**
    *   The application of SMOTE aimed to address the class imbalance, particularly for the 'a-plus' category, which has very few samples (36 in the test set). While SMOTE increased the recall for the minority 'a-plus' class for Naive Bayes (from 0.00 to 0.39), it often came at the cost of precision and overall accuracy for Naive Bayes and Random Forest.
    *   For XGBoost, SMOTE led to a minor accuracy improvement, suggesting it might have helped in better learning the decision boundaries across classes, but the macro average metrics across all models show that the `a-plus` class remains challenging to predict accurately.

**Recommendation:**

XGBoost appears to be the most suitable model for this classification task, offering the best balance of accuracy and generalization. While SMOTE's impact on overall accuracy was marginal or negative for Naive Bayes and Random Forest, the slight improvement for XGBoost suggests that further tuning of SMOTE or XGBoost parameters could potentially yield better results, especially for minority classes if their accurate prediction is a critical goal.